discovery/kubernetes/client_metrics.go

```diff
@@ -0,0 +1,225 @@
+		prometheus.SummaryOpts{
+			Namespace: metricsNamesapce,
+			Name:      "http_request_duration_seconds",
+			Help:      "Request latency of the client-go on Kubernetes API.",
+		},
+		[]string{"verb", "endpoint"},
+	)
+
+	// Definition of metrics for client-go cache metrics provider
+	clientGoCacheListTotalMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "cache_list_total",
+			Help:      "Total number of client-go Kubernets API calls.",
+		},
+	)
+	clientGoCacheListDurationMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_duration_seconds",
+			Help:      "Duration of a Kubernetes API call in seconds.",
+		},
+	)
+	clientGoCacheItemsInListCountMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_items_count",
+			Help:      "Count of items in a list from the Kubernetes API.",
+		},
+	)
+	clientGoCacheWatchesCountMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "watches_total",
+			Help:      "Count of watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheShortWatchesCountMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "short_watches_total",
+			Help:      "Count of short watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheWatchesDurationMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "watch_duration_seconds",
+			Help:      "Duration of watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheItemsInWatchesCountMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "watch_items",
+			Help:      "Number of items in watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheLastResourceVersionMetric = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "last_resource_version",
+			Help:      "Last resource version from the Kubernetes API.",
+		},
+	)
+
+	// Definition of metrics for client-go workflow metrics provider
+	clientGoWorkqueueDepthMetric = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: workqueueMetricsNamespace,
+			Name:      "depth",
+			Help:      "Current depth of a client-go workqueue.",
+		},
+	)
+	clientGoWorkqueueAddsMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: workqueueMetricsNamespace,
+			Name:      "adds_total",
+			Help:      "Total count of adds handeled by a client-go workqueue.",
+		},
+	)
+	clientGoWorkqueueRetriesMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: workqueueMetricsNamespace,
+			Name:      "retries_total",
+			Help:      "Total count of retries in a client-go workqueue.",
+		},
+	)
+	clientGoWorkqueueLatencyMetric = prometheus.NewSummary(
```

Comment:
The `workqueue` package uses microseconds instead of seconds (sigh). We can either change the metric to `latency_microseconds` or (better) have a wrapper around the metric.

```go
clientGoWorkqueueLatencyMetric = prometheus.NewHistogram(...)

...

func (clientGoWorkqueueMetricsProvider) NewLatencyMetric(o prometheus.Observer) workqueue.SummaryMetric {                        
    // Convert microseconds to seconds for consistency across metrics.                                                                     
    return prometheus.ObserverFunc(func(v float64) {
        o.Observe(v / 1e6)
    })
}
```

Reply:
Oh I didn't even notice, thanks :/

Ok I'll create a wrapper, but `NewLatencyMetric` needs `string` as an argument and needs to return an implementation of `workqueue.SummaryMetric` having the `Observe` method.

If I understand correctly your solution returns just the function to observe. I'll have to implement the `Observer` interface which will wrap the `clientGoWorkqueueLatencyMetric` I suppouse.

Nevertheless I now noticed I'm tossing away the queue `name`. I need to add it as a label since right now different queues would override it's gauge metric `NewDepthMetric`.

Reply:
> If I understand correctly your solution returns just the function to observe.

Sorry I messed up my code. I've edited it and should be correct now (albeit the unused `name` parameter).

> I'm tossing away the queue name. I need to add it as a label 

Right and for all queue metrics.